src/toolchain.ml

```diff
@@ -416,23 +418,120 @@ let get_command toolchain bin args : Cmd.t =
+    | None ->
+      show_message `Warn "No installaction method for the current toolchain";
+      resolve ()
+    | Some cmd ->
+      let child_process = Cmd.run ~cwd cmd in
+      let _ =
+        (CancellationToken.onCancellationRequested token) ~listener:(fun _ ->
+            let _ = ChildProcess.kill child_process "SIGTERM" in
+            ())
+      in
+      ChildProcess.on child_process "close" (fun _ ->
+          show_message `Info
+            "The platform tools have been successfully installed";
+          resolve ());
+      ChildProcess.on child_process "error" (fun err ->
+          show_message `Error "The installation of the Platform tools failed";
+          reject err);
+      () )
+
+let install_tools ~progress:_ ~token:_ toolchain tools =
+  let open Promise.Syntax in
+  let workspace_root = workspace_root () in
+  match workspace_root with
+  | None ->
+    show_message `Warn "Multiple workspaces are not supported at the moment";
+    Promise.return ()
+  | Some cwd -> (
+    match get_install_command toolchain tools with
+    | None ->
+      show_message `Warn "No installaction method for the current toolchain";
+      Promise.return ()
+    | Some cmd ->
+      let+ _result = Cmd.output ~cwd cmd in
+      () )
+
+let suggest_to_install_tools toolchain tools =
+  let open Promise.Syntax in
+  let select_pm_button_text = "Install tools" in
+  let missing_str = String.concat ~sep:", " tools in
+  let+ selection =
+    Window.showInformationMessage
+      ~message:
+        (Printf.sprintf
+           "OCaml Platform tools are missing in your sandbox. Do you want to \
+            install them?\n\
+            Missing: %s" missing_str)
+      ~choices:[ (select_pm_button_text, ()) ]
+      ()
+  in
+  let progress_options =
+    ProgressOptions.create ~location:(`ProgressLocation Window)
+      ~title:"Installing Platform Tools" ~cancellable:false ()
+  in
+  let task = install_tools toolchain tools in
+  Option.iter selection ~f:(Window.withProgress ~options:progress_options ~task)
+
+let missing_command_opt ~binary toolchain cmd =
   let open Promise.Syntax in
   let+ output =
     let open Promise.Result.Syntax in
     let* () = setup_toolchain toolchain in
-    let args = [ "--version" ] in
-    let* command = Cmd.check (get_lsp_command toolchain ~args) in
+    let* command = Cmd.check cmd in
     Cmd.output command
   in
   match output with
-  | Ok _ -> Ok ()
-  | Error msg ->
-    (* TODO: if ocamllsp not found, suggest to install it on press of a button;
-       consider checking and suggesting installation for other tools: formatter, etc. *)
-    Error (Printf.sprintf "Toolchain initialisation failed: %s" msg)
+  | Ok _ -> None
+  | Error _ -> Some binary
+
+let run_setup toolchain =
+  let open Promise.Syntax in
+  let* missing_tools =
+    [ ("ocaml-lsp-server", get_lsp_command toolchain ~args:[ "--version" ])
+    ; ("ocamlformat", get_ocamlformat_command toolchain ~args:[ "--version" ])
```

Comment:
Just a heads up regarding formatters and formatting:

You may have noticed that formatting is done by ocaml-lsp, not the extension. Moreover, `ocamllsp` invokes either `ocamlformat` or `refmt` based on the file that needs to be formatted. So it may make sense to handle logic for detecting whether a formatter is installed on the ocaml-lsp side. The lang server could send a notification if it can't find the required tool.

If you choose to do it on the extension side, keep in mind that a user may only need refmt installed or ocamlformat or both, so it's not straightforward.

Reply:
Sorry my TODO comment was misleading

Reply:
Makes sense - do you know if there's a specific notification for that purpose? I've tried to see if LSP had requests/notifications for this kind of things, but did not find anything.